### PR TITLE
DATACASS-529 - Allow slice queries using reactive repositories.

### PR DESCRIPTION
**Pull request by hleb-albau**
Reactive repositories now support methods returning `Mono<Slice<T>>`.
```
Mono<Slice<User>> findByLastname(String firstname, Pageable pageable);
```

Related ticket: DATACASS-529.

### CODE REVIEW

**Comment by mp911de** [comment]
Thanks a lot, we'll take a look as soon as we have time.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/QueryUtils.java [hunk 1]
```diff
@@ -165,4 +169,23 @@ static Delete createDeleteQuery(String tableName, Object objectToDelete, QueryOp
...
+
+	/**
+	 * Read a {@link Mono<Slice>} of data from the {@link ReactiveResultSet} for a {@link Pageable}.
+	 *
+	 * @param resultSet must not be {@literal null}.
+	 * @param fetchSize must not be {@literal null}.
+	 * @param rowMapper must not be {@literal null}.
+	 * @return the resulting {@link Slice}.
+	 */
+	static <T> Mono<Slice<T>> readSlice(ReactiveResultSet resultSet, Integer fetchSize, Function<Row, T> rowMapper) {
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/QueryUtils.java]: We can't introduce dependencies to reactive types here as we want the module to run without reactive dependencies as well.

We also need to be careful about fetching/reading progress. If the fetch size differs from the actual number of rows returned for a page, we might request the next page without intending to do so. I think we need another method on `ReactiveResultSet` that gives us a `Flux` without transparent next page fetching.

File: spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/ReactiveCassandraOperations.java [hunk 2]
```diff
@@ -97,6 +99,19 @@
...
 	<T> Flux<T> select(Statement statement, Class<T> entityClass) throws DataAccessException;
 
+	/**
+	 * Execute a {@code SELECT} query with paging and convert the result set to a {@link Slice} of entities.
+	 *
+	 * A sliced query translates the effective {@link Statement#getFetchSize() fetch size} to the page size.
+	 *
+	 * @param statement the CQL statement, must not be {@literal null}.
+	 * @param entityClass The entity type must not be {@literal null}.
+	 * @return the result object returned by the action or {@link Mono#empty()}
```
Comment (mp911de) [inline spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/ReactiveCassandraOperations.java]: In this case, we're mixing paradigms. We should return an empty `Slice` (see `Page.empty()`) as the query terminates without rows and the query result is the container for rows, not the rows themselves.

**Comment by hleb-albau** [comment]
Hi @mp911de,
I have no free time in near future, so if you can finish this task, it would be nice. Thanks

**Comment by mp911de** [comment]
Cool, thanks for your timely update. We'll take care of this one.

**Comment by eximius313** [comment]
@mp911de do you have any plans when you could handle this?

**Comment by mp911de** [comment]
This pull request needs some polishing before we can merge it. We intend to ship the change with the next release candidate.

**Comment by eximius313** [comment]
So `2.1.0.RC1`?

**Comment by mp911de** [comment]
@lukasz-gosiewski Please file a new ticket in [Spring Data Cassandra](https://jira.spring.io/browse/DATACASS). We'll see what we can do for you.

**Comment by mp911de** [comment]
That's merged and polished now. Thanks a lot!

**Pull request closed without merging**